pkg/storage/redis/redis.go

```diff
@@ -0,0 +1,241 @@
+// Package redis is the redis database, implements storage.Backend interface.
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/gomodule/redigo/redis"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	internal_databroker "github.com/pomerium/pomerium/internal/databroker"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/storage"
+)
+
+var _ storage.Backend = (*DB)(nil)
+
+// DB wraps redis conn to interact with redis server.
+type DB struct {
+	pool                   *redis.Pool
+	deletePermanentlyAfter int64
+	recordType             string
+	lastVersion            uint64
+	versionSet             string
+	deletedSet             string
+}
+
+// New returns new DB instance.
+func New(address, recordType string) (*DB, error) {
+	db := &DB{
+		pool: &redis.Pool{
+			Wait: true,
+			DialContext: func(ctx context.Context) (redis.Conn, error) {
+				ctx, cancelFn := context.WithTimeout(ctx, 5*time.Second)
+				defer cancelFn()
+				c, err := redis.DialContext(ctx, "tcp", address)
+				if err != nil {
+					return nil, fmt.Errorf(`redis.DialURL(): %w`, err)
+				}
+				return c, nil
+			},
+			TestOnBorrow: func(c redis.Conn, t time.Time) error {
+				if time.Since(t) < time.Minute {
+					return nil
+				}
+				_, err := c.Do("PING")
+				if err != nil {
+					return fmt.Errorf(`c.Do("PING"): %w`, err)
+				}
+				return nil
+			},
+		},
+		deletePermanentlyAfter: int64(internal_databroker.DefaultDeletePermanentlyAfter.Seconds()),
+		recordType:             recordType,
+		versionSet:             "version_set",
+		deletedSet:             "deleted_set",
+	}
+	return db, nil
+}
+
+// Put sets new record for given id with input data.
+func (db *DB) Put(ctx context.Context, id string, data *anypb.Any) error {
+	c := db.pool.Get()
+	defer c.Close()
+	record := db.Get(ctx, id)
+	if record == nil {
+		record = new(databroker.Record)
+		record.CreatedAt = ptypes.TimestampNow()
+	}
+
+	record.Data = data
+	record.ModifiedAt = ptypes.TimestampNow()
+	record.Type = db.recordType
+	record.Id = id
+	record.Version = fmt.Sprintf("%012X", atomic.AddUint64(&db.lastVersion, 1))
+	b, err := proto.Marshal(record)
+	if err != nil {
+		return err
+	}
+	cmds := map[string][]interface{}{
+		"MULTI": nil,
+		"HSET":  {db.recordType, id, string(b)},
+		"ZADD":  {db.versionSet, db.lastVersion, id},
+	}
+	if _, err := db.tx(c, cmds); err != nil {
```

Comment:
Map loops in go are non-deterministic. I think the `MULTI` needs to happen first, so you'll need to use a slice instead.

Reply:
Ah yes, my bad.

Reply:
I don't think these changes fixed the issue.

Reply:
`map[string][]interface{}` was changed to `[]map[string][]interface{}` but the loop over the map is still non-deterministic. `MULTI` can be called after `HSET`.

Reply:
LOL, sorry, that the construction is wrong, fixed.